codegen/gce/src/expressions.rs

```diff
@@ -0,0 +1,364 @@
+use super::error::ConstraintEvaluationError;
+use super::{
+    utils::{
+        get_constant_index_by_matrix_access, get_constant_index_by_name,
+        get_constant_index_by_value, get_constant_index_by_vector_access, get_public_input_index,
+        get_random_value_index,
+    },
+    ExpressionJson, ExpressionOperation, NodeReference, NodeType,
+};
+use ir::{
+    constraints::{ConstantValue, Operation},
+    AirIR, NodeIndex,
+};
+use std::collections::BTreeMap;
+
+const MAIN_TRACE_SEGMENT_INDEX: u8 = 0;
+
+pub struct ExpressionsHandler<'a> {
+    ir: &'a AirIR,
+    constants: &'a [u64],
+    // maps indexes in Node vector in AlgebraicGraph and in `expressions` JSON array
+    expressions_map: &'a mut BTreeMap<usize, usize>,
+}
```

Comment:
As noted, the `expressions_map` can be instantiated in the `new` method, so it no longer needs to be marked as a mut borrow with a lifetime.

I think we should also add the following. As mentioned, we can declare them in `new` and then populate them in a `build` method.

```
    expressions: Vec<ExpressionJson>,
    outputs: Vec<usize>,
```

I would also probably remove the `AirIR` and `constants` borrows here and just pass them to the `build` method.

For the `build` method, we could do something like the following:

```
    pub fn build(
        &mut self,
        ir: &AirIR,
        constants: &[u64],
    ) -> Result<(), ConstraintEvaluationError> {
        self.build_expressions(ir, constants)?;
        self.build_outputs(ir)?;
        Ok(())
    }
```